README.md
---
page_type: sample
languages:
- powershell
products:
- azure
description: "Automates top asks using PowerShell for Azure Backup archive feature"
---

# Automate top asks using PowerShell for Azure Backup

Automate archive move using [PowerShell for Azure Backup](https://docs.microsoft.com/en-us/azure/backup/archive-tier-support)

## Features

Runbooks for Archive move

## Step1
Create an Automation resource with “Run As” account

## Step2
Import modules from Gallery in the Automation resource

Import the following modules from the Modules gallery:
1. Az.RecoveryServices 4.0.0-preview 

## Step3
Create PowerShell Runbooks in the Automation Resource. You can create multiple Runbooks based on which set of RPs you want to move

## Step4
Edit the Runbook and write script to choose BackupItem for archive move. You can create scripts that suit your requirements.
- Save the script
- Test the script using “Test Pane”
- Publish the Runbook

Scripts for different usages are provided in the repository

## Step5
Schedule the Runbook. While scheduling the Runbook, you can pass on the parameters required for the PowerShell Script. You can use '-?' to see the needed parameters for each script.


CHANGELOG.md
## [ArchiveFeatureSupport] Changelog

*Features*
* Move archivable RecoveryPoints for SQL
* Move recommended Recovery Points for IaasVM
* View archivable Recovery Points in a given time range


ArchiveFeatureSupport/moveRecommendedRPsForIaasVM.ps1
<#
.SYNOPSIS
    Moves recommended RPs for a given IaasVM
.DESCRIPTION
    Moves recommended RPs for a given IaasVM.
    This requires PowerShell 7.0 and Az.RecoveryServices preview module installed 
#>

param( 
    [Parameter(Mandatory=$true, HelpMessage="SubscriptionId or SubscriptionName ")] 
    [string] $Subscription, 

    [Parameter(Mandatory=$true)] 
    [string] $ResourceGroupName,

    [Parameter(Mandatory=$true)] 
    [string] $VaultName,
    
    [Parameter(Mandatory=$true, HelpMessage="Name of Backup Item")] 
    [string] $BackupItemName
)

function script:TraceMessage([string] $message, [string] $color="Yellow")
{
    Write-Host "`n$message" -ForegroundColor $color
}

try
{
    Set-AzContext -Subscription $Subscription | Out-Null
}
catch
{
    Add-AzAccount
    Set-AzContext -Subscription $Subscription | Out-Null
}

#fetch recovery services vault  
$vault =  Get-AzRecoveryServicesVault -ResourceGroupName $ResourceGroupName -Name $VaultName

# fetch vm backup items
$vmItems = Get-AzRecoveryServicesBackupItem -BackupManagementType "AzureVM" `
    -WorkloadType "AzureVM" -VaultId $vault.ID | Where-Object { $_.Name -eq $BackupItemName}

$result = @()    
# for each vm item - move all recommended RPs to Archive
foreach ($vmItem in $vmItems){
    $recRPList = Get-AzRecoveryServicesBackupRecommendedArchivableRPGroup -Item $vmItem -VaultId $vault.ID

    foreach ($rp in $recRPList){
        $job = Move-AzRecoveryServicesBackupRecoveryPoint -RecoveryPoint $rp `
        -SourceTier $rp.RecoveryPointTier -DestinationTier VaultArchive -VaultId $vault.ID

        $result = $result + $job
    }
}

Write-Output($result)

ArchiveFeatureSupport/moveArchivableRecoveryPointsForSQL.ps1
<#
.SYNOPSIS
    Moves archivable RPs for a given SQL backup item
.DESCRIPTION
    Moves archivable RPs for a given SQL backup item to the VaultArchive tier. 
    By default this script moves archivable RPs for last 2 years.
    This requires PowerShell 7.0 and Az.RecoveryServices preview module installed 
#>

param( 
    [Parameter(Mandatory=$true)] 
    [string] $Subscription,

    [Parameter(Mandatory=$true)] 
    [string] $ResourceGroupName,

    [Parameter(Mandatory=$true)] 
    [string] $VaultName,

    [Parameter(Mandatory=$false, HelpMessage="Start Date in Utc")] 
    [System.DateTime] $StartDate = (Get-Date).AddDays(-730).ToUniversalTime(),

    [Parameter(Mandatory=$false, HelpMessage="End Date in Utc")] 
    [System.DateTime] $EndDate = (Get-Date).AddDays(0).ToUniversalTime(),
    
    [Parameter(Mandatory=$true, HelpMessage="Name of Backup Item")] 
    [string] $BackupItemName
)

function script:TraceMessage([string] $message, [string] $color="Yellow")
{
    Write-Host "`n$message" -ForegroundColor $color
}

try
{
    Set-AzContext -Subscription $Subscription | Out-Null
}
catch
{
    Add-AzAccount
    Set-AzContext -Subscription $Subscription | Out-Null
}

#fetch recovery services vault  
$vault =  Get-AzRecoveryServicesVault -ResourceGroupName $ResourceGroupName -Name $VaultName
    
# fetch SQL backup item within the vault
$sqlItems = Get-AzRecoveryServicesBackupItem -BackupManagementType "AzureWorkload" `
    -WorkloadType "MSSQL" -VaultId $vault.ID | Where-Object { $_.Name -eq $BackupItemName}

# for each sql item - move all move-ready recovery points (wihin given time range) to Archive
foreach ($sqlItem in $sqlItems){

    $EndDate1 = $EndDate
    while ($EndDate1 -ge $StartDate) {
        $timeDiff = ($EndDate1 - $StartDate)
    
        if($timeDiff.Days -ge 30){
            $StartDate1 = $EndDate1.AddDays(-30).ToUniversalTime()
        }
        else {
            $StartDate1 = $StartDate
        }

        $archivableSQLRPs = Get-AzRecoveryServicesBackupRecoveryPoint -Item $sqlItem `
        -StartDate $StartDate1 -EndDate $EndDate1 -VaultId $vault.ID -IsReadyForMove $true `
        -TargetTier VaultArchive

        $allRecoveryPoints = $allRecoveryPoints + $archivableSQLRPs 
        
        $EndDate1 = $EndDate1.AddDays(-30).ToUniversalTime() 
    }                     
}

# for each sql item - move all move-ready recovery points (wihin given time range) to Archive
$result = @()
foreach ($rp in $allRecoveryPoints){
    $job = Move-AzRecoveryServicesBackupRecoveryPoint -RecoveryPoint $rp `
        -SourceTier $rp.RecoveryPointTier -DestinationTier VaultArchive -VaultId $vault.ID
    
    $result = $result + $job
}

Write-Output($result)

ArchiveFeatureSupport/viewArchivableRPs.ps1
<#
.SYNOPSIS
    View Archivable Recovery Points
.DESCRIPTION
    View Archivable Recovery Points.
    By default this script list RPs for last 2 years.
    This requires PowerShell 7.0 and Az.RecoveryServices preview module installed 
#>

param( 
    [Parameter(Mandatory=$true)] 
    [string] $Subscription,

    [Parameter(Mandatory=$true)] 
    [string] $ResourceGroupName,

    [Parameter(Mandatory=$true)] 
    [string] $VaultName,

    [Parameter(Mandatory=$true, HelpMessage="Valid values: AzureVM, MSSQL")] 
    [string] $ItemType,

    [Parameter(Mandatory=$false, HelpMessage="Start Date in Utc")] 
    [System.DateTime] $StartDate = (Get-Date).AddDays(-730).ToUniversalTime(),

    [Parameter(Mandatory=$false, HelpMessage="End Date in Utc")] 
    [System.DateTime] $EndDate = (Get-Date).AddDays(0).ToUniversalTime(),
    
    [Parameter(Mandatory=$true, HelpMessage="Name of Backup Item")] 
    [string] $BackupItemName
)

function script:TraceMessage([string] $message, [string] $color="Yellow")
{
    Write-Host "`n$message" -ForegroundColor $color
}

try
{
    Set-AzContext -Subscription $Subscription | Out-Null
}
catch
{
    Add-AzAccount
    Set-AzContext -Subscription $Subscription | Out-Null
}

#fetch recovery services vault  
$vault =  Get-AzRecoveryServicesVault -ResourceGroupName $ResourceGroupName -Name  $VaultName

# Command Output 
    
if($ItemType -eq "AzureVM"){
    # fetch vm backup items
    $vmItems = Get-AzRecoveryServicesBackupItem -BackupManagementType "AzureVM" `
        -WorkloadType "AzureVM" -VaultId $vault.ID | Where-Object { $_.Name -eq $BackupItemName}

    # for each vm item - move all recommended RPs to Archive
    foreach ($vmItem in $vmItems){

        $EndDate1 = $EndDate            
        while ($EndDate1 -ge $StartDate) {
            $timeDiff = ($EndDate1 - $StartDate)
        
            if($timeDiff.Days -ge 30){
                $StartDate1 = $EndDate1.AddDays(-30).ToUniversalTime()
            }
            else {
                $StartDate1 = $StartDate
            }

            $archivableVMRPs = Get-AzRecoveryServicesBackupRecoveryPoint -Item $vmItem `
            -StartDate $StartDate1 -EndDate $EndDate1 -VaultId $vault.ID -IsReadyForMove $true `
            -TargetTier VaultArchive

            # Write-Host ("Archivable Recovery Points under AzureVM " + $vmItem.Name)
            $allRecoveryPoints = $allRecoveryPoints + $archivableVMRPs                       

            $EndDate1 = $EndDate1.AddDays(-30).ToUniversalTime() 
        }            
    }
}
elseif ($ItemType -eq "MSSQL") {
    # fetch all SQL backup items within the vault
    $sqlItems = Get-AzRecoveryServicesBackupItem -BackupManagementType "AzureWorkload" `
        -WorkloadType "MSSQL" -VaultId $vault.ID | Where-Object { $_.Name -eq $BackupItemName}

    # for each sql item - move all move-ready recovery points (wihin given time range) to Archive
    foreach ($sqlItem in $sqlItems){

        $EndDate1 = $EndDate
        while ($EndDate1 -ge $StartDate) {
            $timeDiff = ($EndDate1 - $StartDate)
        
            if($timeDiff.Days -ge 30){
                $StartDate1 = $EndDate1.AddDays(-30).ToUniversalTime()
            }
            else {
                $StartDate1 = $StartDate
            }

            $archivableSQLRPs = Get-AzRecoveryServicesBackupRecoveryPoint -Item $sqlItem `
            -StartDate $StartDate1 -EndDate $EndDate1 -VaultId $vault.ID -IsReadyForMove $true `
            -TargetTier VaultArchive
            
            $allRecoveryPoints = $allRecoveryPoints + $archivableSQLRPs 
            
            $EndDate1 = $EndDate1.AddDays(-30).ToUniversalTime() 
        }                     
    }    
}
else {
    Write-Error "Invalid ItemType. Valid values: AzureVM, MSSQL"
}

Write-Output ($allRecoveryPoints)